### PR TITLE
Istio version checking

### PR DESCRIPTION
**Pull request by werdes72**
See: #124

### CODE REVIEW

File: internal/reconciliations/istio/configuration.go [hunk 1]
```diff
@@ -92,3 +99,51 @@ func UpdateLastAppliedConfiguration(cr operatorv1alpha1.Istio, istioTag string)
...
+	if !amongOneMinor(currentIstioVersion, targetIstioVersion) {
+		return fmt.Errorf("target Istio version (%s) is higher than current Istio version (%s) - the difference between versions exceed one minor version", targetIstioVersion.ver.String(), currentIstioVersion.ver.String())
+	}
+
+	return nil
+}
+
+type helperVersion struct {
+	ver semver.Version
+}
```
Comment (barchw) [inline internal/reconciliations/istio/configuration.go:129]:
```suggestion
type helperVersion semver.Version
```

File: internal/reconciliations/istio/configuration.go [hunk 2]
```diff
@@ -92,3 +99,51 @@ func UpdateLastAppliedConfiguration(cr operatorv1alpha1.Istio, istioTag string)
...
+
+func newHelperVersionFrom(versionInString string) (helperVersion, error) {
+	version, err := semver.NewVersion(versionInString)
+	if err != nil {
+		return helperVersion{}, err
+	}
+	return helperVersion{ver: *version}, err
```
Comment (barchw) [inline internal/reconciliations/istio/configuration.go:136]: We should check for nil

File: internal/reconciliations/istio/configuration.go [hunk 3]
```diff
@@ -92,3 +99,51 @@ func UpdateLastAppliedConfiguration(cr operatorv1alpha1.Istio, istioTag string)
...
+}
+
+func isDowngrade(current, target helperVersion) bool {
+	return current.ver.Major >= target.ver.Major && current.ver.Minor >= target.ver.Minor && current.ver.Patch > target.ver.Patch
```
Comment (barchw) [inline internal/reconciliations/istio/configuration.go:140]:
```suggestion
	return target.ver.LessThan(current.ver)
```

File: internal/reconciliations/istio/configuration.go [hunk 4]
```diff
@@ -92,3 +99,51 @@ func UpdateLastAppliedConfiguration(cr operatorv1alpha1.Istio, istioTag string)
...
+}
+
+func amongOneMajor(first, second helperVersion) bool {
```
Comment (barchw) [inline internal/reconciliations/istio/configuration.go:143]:
```suggestion
func sameMajorVersion(first, second helperVersion) bool {
```

File: internal/reconciliations/istio/configuration.go [hunk 5]
```diff
@@ -92,3 +99,51 @@ func UpdateLastAppliedConfiguration(cr operatorv1alpha1.Istio, istioTag string)
...
+	if err != nil {
+		return err
+	}
+
+	if isDowngrade(currentIstioVersion, targetIstioVersion) {
+		return fmt.Errorf("target Istio version (%s) is lower than current version (%s) - downgrade not supported",
+			targetIstioVersion.ver.String(), currentIstioVersion.ver.String())
+	}
+	if !amongOneMajor(currentIstioVersion, targetIstioVersion) {
+		return fmt.Errorf("target Istio version (%s) is higher than current Istio version (%s) - major version upgrade is not supported", targetIstioVersion.ver.String(), currentIstioVersion.ver.String())
```
Comment (barchw) [inline internal/reconciliations/istio/configuration.go:118]:
```suggestion
		return fmt.Errorf("target Istio version (%s) is different than current Istio version (%s) - major version upgrade is not supported", targetIstioVersion.ver.String(), currentIstioVersion.ver.String())
```

**Review by barchw — CHANGES_REQUESTED** [review]

File: internal/reconciliations/istio/configuration.go (hunk 1 above)
Comment (werdes72) [reply internal/reconciliations/istio/configuration.go]: Removed helper type.

**Review by barchw — APPROVED** [review]
/meow

**Pull request merged**